Issue 7: search and filter items in shopping list

NickRoccodev11 opened the pull request:
## Description
 This PR adds two functionalities to locate items:

- [x]  **Search  Input**: 
 >  - Semantic` label `  that adds keyboard focus to input
 > - `button` that conditionally renders based on `searchInput `state
> - `button` contains an `aria-label` attribute for providing context for screen readers

- [x] **Filter Text**: 
> - filter text should match any part of item name
> - Only list items that match `searchInput` text are rendered


<!-- What does this code change? Why did I choose this approach? Did I learn anything worth sharing? Reminder: This will be a publicly facing representation of your work (READ: help you land that sweet dev gig). -->

## Related Issue

closes #7

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->

## Acceptance Criteria

- [ ] A form is added to the top of the List view, above the shopping list
- [ ]  The form includes the following elements
> - A text field (with semantic <label>!) which narrows down the list as the user types
> - When there’s text in the field, some kind of button (e.g., an X) to clear the field. When the field is cleared, the list is reset to its unfiltered state.

<!-- Include AC from the Github issue -->

## Type of Changes
| Type |                        |
| ---- | ---------------------- |
| ✓    | :sparkles: Enhancement       |
| ✓    | :link: Accessibility |

## Updates

### Before

https://github.com/user-attachments/assets/2e303632-8dc4-429f-a714-65e521876c9e

<!-- If UI feature, take provide screenshots -->

### After

https://github.com/user-attachments/assets/aa67e1c4-92bd-4824-9789-572d11e8310b


<!-- If UI feature, take provide screenshots -->

## Testing Steps / QA Criteria

- [ ] Sign in using your Google account.
- [ ] Navigate to Home page and click on `test-list`.
- [ ] Navigate to List page and click on 'Search Items' label to focus to the input.
- [ ] Inside the input, begin typing an item name. 
- [ ] The displayed list should filter based on your input.
- [ ] To clear the input, click the "x" button.


<!-- Provide steps the other cohort members and mentors need to follow to properly test your additions. -->

NickRoccodev11 committed:
import useState
1 of 3

NickRoccodev11 committed:
Add input and state for search functionality

 2 of 3, button conditionally renders based on state

dterceroparker committed:
Implement filtered functionality

3 of 3

shuveksha-tuladhar reviewed (approved):
Looks great!

stacy-tech reviewed (approved):
Well done guys!

dterceroparker committed:
add aria-label attribute to x button with clear search input text for accessibility purposes

jtkabenni reviewed (approved):
Looks great! Your filtering logic here is perfectly fine but one thing that can be considered is to bring out the filtering into a separate function, and then in the jsx just mapping over the filtered list. 
A few reasons this may be considered - 1. currently filtering will run every time the component is rendered and 2. bringing out the function may make things more testable and reusable! Your logic here is simple so this may be unnecessary, but just something to think about.

NickRoccodev11 committed:
Extract filter method from JSX and make into separate function

update aria-label, update text from 'no items available' to 'no items found'

NickRoccodev11 commented:
> Looks great! Your filtering logic here is perfectly fine but one thing that can be considered is to bring out the filtering into a separate function, and then in the jsx just mapping over the filtered list. A few reasons this may be considered - 1. currently filtering will run every time the component is rendered and 2. bringing out the function may make things more testable and reusable! Your logic here is simple so this may be unnecessary, but just something to think about.

- Thank you for your feedback! We met and updated the code accordingly, but would love to go over some of the details tomorrow. We also updated aria-label per EJ's suggestion. - Doribel and Nick
